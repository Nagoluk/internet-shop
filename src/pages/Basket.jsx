import React from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {getBasket, getTotalCount} from '../store/selectors/basket-selector';
import {PageHeader} from '../components/pageHeader';
import styled from 'styled-components';
import {changeAmountAC} from '../store/reducers/basket-reducer';
import {Button} from '../components/ui/button';
import {NavLink} from 'react-router-dom';
import {FormattedMessage} from 'react-intl';

export const BasketWrap = styled.div`
   max-width: 1200px;
   margin: 0 auto;
`

const BasketItemStyled = styled.div`
    background: ${props => props.theme.backgroundForComponents};
    margin-top: 10px;
    padding: 10px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    
    & input {
        width: 30px;
    }
`

export const OrderButton = styled(Button)`
    font-size: 2rem;
    display: block;
    margin: 10px auto;
    
`

const BasketItem = React.memo( ({_id, name, logo, price, amount, dispatch}) => {
    return <BasketItemStyled>
            <img src={logo} alt={name}/>
            <h3>{name}</h3>
            <div>
                <label htmlFor="amount"><FormattedMessage id={'count'}/> </label>
                <input type="number" id={'amount'} value={amount} onChange={event => {
                    dispatch(changeAmountAC(Number(event.target.value), _id))
                }}/>
            </div>
            <p>
                <FormattedMessage id={'total'}/>: {price * amount} $
            </p>
           </BasketItemStyled>
})

export const Basket = () => {
    const basketList = useSelector(getBasket)
    const totalCount = useSelector(getTotalCount)
    const dispatch = useDispatch()

    return (<>
                <PageHeader title={<FormattedMessage id={'basket'}/>}>
                    <h3><FormattedMessage id={'total'}/>: {totalCount} $</h3>
                </PageHeader>
                <BasketWrap>
                {basketList.map(item => <BasketItem key={item._id} {...item} dispatch={dispatch}/>)}

                {basketList.size === 0 && <NavLink to={'/'}><p><FormattedMessage id={'goToCatalog'}/></p></NavLink>}
                {basketList.size !== 0 && <NavLink to={'/order'}><OrderButton><FormattedMessage id={'makeOrder'}/></OrderButton></NavLink>}
                </BasketWrap>
            </>
    );
};


