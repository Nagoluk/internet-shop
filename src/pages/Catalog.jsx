import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  getCatalogFetching,
  getGoods,
} from "../store/selectors/catalog-selector";
import { PageHeader } from "../components/pageHeader";
import styled from "styled-components";
import { fetchCatalog } from "../store/reducers/catalog";
import { Button } from "../components/ui/button";
import { NavLink } from "react-router-dom";
import { addToBasketAC } from "../store/reducers/basket-reducer";
import { FormattedMessage } from "react-intl";

const GoodsWrap = styled.div`
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  max-width: 1200px;
  margin: 0 auto;
  padding-top: 5px;
`;

const Good = styled.div`
  background: ${(props) => props.theme.backgroundForComponents};
  padding: 10px;
  width: 200px;
  text-align: center;
  margin: 5px;
  border-radius: 5px;
`;

const Price = styled.p`
  margin-bottom: 5px;
  font-size: 0.8rem;
`;

const ToBasket = styled.div`
  & > a {
    color: ${(props) => props.theme.text};
    font-weight: 600;
  }
`;

const GoodItem = React.memo(({ _id, name, logo, price, dispatch }) => {
  return (
    <Good data-qa={_id + "=good"}>
      <img src={logo} alt={name} />
      <div>
        <h3>{name}</h3>
        <Price>{price}$</Price>
        <div>
          <Button
            onClick={() => {
              dispatch(addToBasketAC({ _id, name, logo, price }));
              alert("Added to basket");
            }}
          >
            <FormattedMessage id={"buy"} />
          </Button>
        </div>
      </div>
    </Good>
  );
});

export const Catalog = () => {
  const goods = useSelector(getGoods);
  const isFetch = useSelector(getCatalogFetching);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCatalog());
  }, []);

  return (
    <>
      <PageHeader title={<FormattedMessage id={"catalog"} />}>
        <ToBasket>
          <NavLink to={"basket"}>
            <FormattedMessage id={"basket"} />
          </NavLink>
        </ToBasket>
      </PageHeader>
      {!isFetch && (
        <GoodsWrap>
          {goods.map((good, index) => (
            <GoodItem key={good._id} {...good} dispatch={dispatch} />
          ))}
        </GoodsWrap>
      )}

      {isFetch && <div>Fetching catalog....</div>}
    </>
  );
};
