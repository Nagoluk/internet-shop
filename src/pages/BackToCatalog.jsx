import React from 'react';
import {PageHeader} from '../components/pageHeader';
import styled from 'styled-components';
import {OrderButton} from './Basket';
import {useHistory} from 'react-router-dom'
import {FormattedMessage} from 'react-intl';


const BigText = styled.h1`
    font-size: 10rem;
    text-align: center;
`
const Back = styled(OrderButton)`
    margin-top: 100px;
`

export const BackToCatalog = () => {
    const history = useHistory()
    return (<>
                <PageHeader title={<FormattedMessage id={'thankYou'}/>}/>
                <BigText><FormattedMessage id={'thankYou'}/> !</BigText>
                <Back onClick={() => history.push('/')}><FormattedMessage id={'goToCatalog'}/></Back>
            </>
    );
};


