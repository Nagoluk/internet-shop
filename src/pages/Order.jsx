import React, {useEffect} from 'react';
import {PageHeader} from '../components/pageHeader';
import {Button} from '../components/ui/button';
import styled from 'styled-components';
import {BasketWrap} from './Basket';
import {reduxForm, Field} from 'redux-form';
import {required, validateEmail} from '../utils/redux-form-validators';
import {InputField} from '../components/form/input';
import {useDispatch, useSelector} from 'react-redux';
import {sendOrderThunk, setRedirectToFinish} from '../store/reducers/basket-reducer';
import {getBasketLength, getRedirectToFinish} from '../store/selectors/basket-selector';
import {NavLink, useHistory} from 'react-router-dom';
import {FormattedMessage} from 'react-intl';

const FormWrap = styled(BasketWrap)``

const FormStyled = styled.form`
    background: ${props => props.theme.backgroundForComponents};
    padding: 20px 10px;
    margin-top: 10px;
`

const Label = styled.label`
    font-size: 0.9rem;
    margin-right: 5px;
    
    & .required {
        color: red;
    }
`

const FormItem = styled.div`
    margin-bottom: 5px;
    max-width: 270px;
    ${props => (props.layout === 'horizontal') ? 'display: flex; justify-content: space-between; align-items: center': ''}
`

const FormTitle = styled.h3`
    margin-bottom: 10px;
`


const Form = ({handleSubmit, submitting, error}) => {
    return (<FormStyled onSubmit={handleSubmit}>
                <FormTitle><FormattedMessage id={'userInfo'}/></FormTitle>
                <FormItem>
                    <Label><FormattedMessage id={'name'}/> <span className={'required'}>*</span></Label>
                    <Field type={'text'} component={InputField} validate={required} name={'name'}/>
                </FormItem>

                <FormItem>
                    <Label>Email <span className={'required'}>*</span></Label>
                    <Field type={'text'}
                           component={InputField}
                           validate={[required, validateEmail]} name={'email'}/>
                </FormItem>

                <FormTitle><FormattedMessage id={'deliverAddress'}/></FormTitle>
                <FormItem layout={'horizontal'}>
                    <Label><FormattedMessage id={'city'}/> <span className={'required'}>*</span></Label>
                    <Field type={'text'} component={InputField} validate={required} name={'city'}/>
                </FormItem>

                <FormItem layout={'horizontal'}>
                    <Label><FormattedMessage id={'postalcode'}/>: <span className={'required'}>*</span></Label>
                    <Field type={'text'} component={InputField} validate={required} name={'code'}/>
                </FormItem>


                <div>
                    <Button type="submit" disabled={submitting}><FormattedMessage id={'submit'}/></Button>
                </div>
    </FormStyled>)
}

let ReduxOrderForm = reduxForm({form: 'order'})(Form)

export const Order = () => {
    const dispatch = useDispatch()
    const history = useHistory()
    const basketLength = useSelector(getBasketLength)
    const redirectToFinish = useSelector(getRedirectToFinish)

    useEffect(() => {
        if(redirectToFinish) history.push('/finish')
        return () => {
            dispatch(setRedirectToFinish(false))
        }
    }, [redirectToFinish, dispatch, history])

    const handleSubmit = (data) => {
        dispatch(sendOrderThunk(data))
    }

    return (<>
                <PageHeader title={'Order'}/>
                {!!basketLength && <FormWrap>
                    <ReduxOrderForm onSubmit={handleSubmit}/>
                </FormWrap>}

            {(!basketLength) && <NavLink to={'/'}>Basket is empty go to catalog</NavLink>}
            </>
    );
};


