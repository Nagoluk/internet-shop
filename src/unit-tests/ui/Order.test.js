import {act, render} from '@testing-library/react'

import React from 'react';
import {UniversalProvider} from '../../index';


import { unmountComponentAtNode } from "react-dom";
import {Order} from '../../pages/Order';


let container = null;
beforeEach(() => {
    // подготавливаем DOM-элемент, куда будем рендерить
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // подчищаем после завершения
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

it("renders <Order/> without crashing", () => {
    act(() => {
        render(<UniversalProvider>
            <Order/>
        </UniversalProvider>, container);
    });
})

it('<Order/> matches snapshot', () => {
    const component = render(<UniversalProvider><Order/></UniversalProvider>, container)

    expect(component.container).toMatchSnapshot()
})