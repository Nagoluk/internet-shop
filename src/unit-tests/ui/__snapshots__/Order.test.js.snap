// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order/> matches snapshot 1`] = `
<div>
  <header
    class="sc-bdnxRM kODaaW"
  >
    <div
      class="sc-gtsrHT bkbeeT"
    >
      <select
        name="lang"
      >
        <option
          value="en-us"
        >
          en
        </option>
        <option
          value="uk-UA"
        >
          ua
        </option>
      </select>
      <select
        name="theme"
      >
        <option
          value="dark"
        >
          dark
        </option>
        <option
          value="light"
        >
          light
        </option>
      </select>
      <h1>
        Order
      </h1>
    </div>
    <div
      class="sc-gtsrHT sc-dlnjwi bkbeeT jFMDrI"
    />
  </header>
  <a
    aria-current="page"
    class="active"
    href="/"
  >
    Basket is empty go to catalog
  </a>
</div>
`;
